### PR TITLE
Add missing operations to ArrayOps

### LINKED ISSUE #567: ArrayOps doesn’t provide a sliding operation
A consequence of that is that the one from `WrappedArray` is picked, but this one returns an `Iterator[WrappedArray[A]]`.

See https://github.com/scalaz/scalaz/pull/1730#discussion_r182999064

### PR DESCRIPTION
**Pull request by szeiger**
plus some related bug fixes

Fixes https://github.com/scala/collection-strawman/issues/567

### CODE REVIEW

**Comment by szeiger** [comment]
The test errors appear to be caused by a binary incompatible partest after bootstrapping, so this needs to be rebased after merging https://github.com/scala/scala/pull/6566

File: src/library/scala/collection/ArrayOps.scala [hunk 1]
```diff
@@ -133,31 +166,37 @@ final class ArrayOps[A](val xs: Array[A]) extends AnyVal {
...
-  def nonEmpty: Boolean = xs.length != 0
+  @`inline` def nonEmpty: Boolean = xs.length != 0
 
   /** Selects the first element of this array.
     *
     *  @return  the first element of this array.
     *  @throws NoSuchElementException if the array is empty.
     */
-  def head: A = xs.apply(0)
+  def head: A = try xs.apply(0) catch { case _: ArrayIndexOutOfBoundsException => throw new NoSuchElementException }
```
Comment (lrytz) [inline src/library/scala/collection/ArrayOps.scala]: Should we pass on the exception message? Or maybe use `initCause`?

Comment (lrytz) [reply src/library/scala/collection/ArrayOps.scala]: Actually I realize that doesn't make sense for head / last, so this looks fine.

Comment (lrytz) [reply src/library/scala/collection/ArrayOps.scala]: `"head/last of empty array"` would make sense though

File: src/library/scala/collection/ArrayOps.scala [hunk 2]
```diff
@@ -721,7 +871,24 @@ final class ArrayOps[A](val xs: Array[A]) extends AnyVal {
...
+  /** Finds the first element of the $coll for which the given partial function is defined, and applies the
+    * partial function to it. */
+  def collectFirst[B : ClassTag](f: PartialFunction[A, B]): Option[B] = {
+    var i = 0
+    var matched = true
+    def d(x: A): B = {
+      matched = false
+      null.asInstanceOf[B]
+    }
+    val b = ArrayBuilder.make[B]()
```
Comment (lrytz) [inline src/library/scala/collection/ArrayOps.scala]: this seems to be unused

File: src/library/scala/collection/ArrayOps.scala [hunk 3]
```diff
@@ -261,6 +334,23 @@ final class ArrayOps[A](val xs: Array[A]) extends AnyVal {
...
+    *  predicate `p` does not cause any side-effects.
+    *
+    *  @param p the test predicate
+    *  @return  a pair consisting of the longest prefix of this array whose
+    *           chars all satisfy `p`, and the rest of this array.
+    */
+  def span(p: A => Boolean): (Array[A], Array[A]) = {
+    val i = indexWhere(x => !p(x))
+    val hi = if(i < 0) xs.length else i
+    val lo = if(i < 0) xs.length else i
```
Comment (lrytz) [inline src/library/scala/collection/ArrayOps.scala]: `hi` and `lo` are identical, that looks suspicious

**Review by lrytz — CHANGES_REQUESTED** [review]

**Review by lrytz — APPROVED** [review]

**Pull request merged**